Support go to declaration for `Mapper#qualifiedByName`

thunderhook opened the pull request:
This pull request closes #107.

What it does:
* Provide code-completion for `@Mapping#qualifiedByName`
* Code-complete "local" named methods (from the current mapper class)
* Code-complete "external" named methods (defined by `@Mapper(uses = AnotherMapper.class)`)
* Filters out named methods that are used for `@BeforeMapping/@AfterMapping`: I just provide named methods containing a return type different than `PsiType.VOID` - maybe this needs to be more precise?

What it does not:
* Provide code-completion of the `@BeanMapping#qualifiedByName`: I tried to implement it, but discarded it due to the complexity of nested qualifiers as seen [here](https://stackoverflow.com/a/68966277/1213544)

almogtavor commented:
@thunderhook Looks awesome thanks!

thunderhook commented:
Since I am not a part of the plugin maintainers, I can't really tell. It depends on the free time of the plugin maintainers (e.g. @filiphr) and if the pull request looks good to them.

If they get time, it usually will be released quickly. 🙂

filiphr commented on src/main/java/org/mapstruct/intellij/util/MapstructAnnotationUtils.java:
```diff
@@ -370,4 +375,41 @@ public static PsiModifierListOwner findMapperConfigReference(PsiAnnotation mappe
+     * @return the classes / interfaces that are defined with the {@code uses} attribute,
+     * or {@code null} if there isn't anything defined
+     */
+    public static List<PsiClass> resolveUsesConfigClasses(PsiAnnotation mapperAnnotation) {
```
Can we also look into the `Mapper#config` class for this? We do something with it in `TargetUtils#isBuilderEnabled` (as a reference example). Would also be good to add a test case for it as well

filiphr commented on src/main/java/org/mapstruct/intellij/codeinsight/references/MapstructMappingQualifiedByNameReference.java:
```diff
@@ -0,0 +1,173 @@
+            mappingMethod );
+
+        return allMethodsFromThisAndReferencedMappers.stream()
+            .filter( MapstructUtil::isNamedMethod )
+            .filter( this::methodHasReturnType )
```
Let's move these checks in the `findAllMethodsFromThisAndReferencedMappers`.  That one is already partially doing this for the named methods in used mappers.

filiphr commented on src/main/java/org/mapstruct/intellij/codeinsight/references/MapstructMappingQualifiedByNameReference.java:
```diff
@@ -0,0 +1,173 @@
+            mappingMethod );
+
+        return allMethodsFromThisAndReferencedMappers.stream()
+            .filter( MapstructUtil::isNamedMethod )
+            .filter( this::methodHasReturnType )
```
Let's move these checks in the `findAllMethodsFromThisAndReferencedMappers`.  That one is already partially doing this for the named methods in used mappers.

filiphr commented on src/main/java/org/mapstruct/intellij/codeinsight/references/MapstructMappingQualifiedByNameReference.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    @NotNull
+    private List<PsiMethod> findAllMethodsFromThisAndReferencedMappers(@NotNull PsiMethod mappingMethod) {
```
I think that it is better if we return a `Stream<PsiMethod>` here

filiphr commented on src/main/java/org/mapstruct/intellij/codeinsight/references/MapstructMappingQualifiedByNameReference.java:
```diff
@@ -0,0 +1,173 @@
+            method,
+            lookupString,
+            lookupString,
+            String.format(
```
I think that for this it would be good to show the method parameters here as well. e.g. `StringMapper#trim(String)`.

filiphr reviewed (changes requested):
This is really great work @thunderhook. The functionality is great. I've added some comments to clean it up a bit.

Please let us know if you don't have the time to finish it.

Btw. sorry for the late reply, but I've been pretty busy lately and haven't had enough time to look into this.

thunderhook commented:
I added the named method resolving of mappers that are referenced by `@MapperConfig`.
I also changed a named method in a test to cover the display of multiple parameters.
Please take a look and let me know if there is something else missing. 🙂

If you have any other ideas for this plugin (with that kind of complexity), please let me know, and I'm willing to help out again. 👍

filiphr commented:
That's a lot for your changes @thunderhook. I pushed one more commit with some small polishing.

>If you have any other ideas for this plugin (with that kind of complexity), please let me know, and I'm willing to help out again.

Thanks for offering your help. I think that you can have a look at the open issues to see what is being requested. If you are interested in something let us know. Also if you have some other ideas that you'd like to see let us know, we are more than happy to accept your contributions.
